token-lending/program/src/processor.rs

```diff
@@ -1016,141 +884,146 @@ fn process_liquidate(
+        msg!("Lending market token program does not match the token program provided");
         return Err(LendingError::InvalidTokenProgram.into());
     }
 
-    let mut obligation = Obligation::unpack(&obligation_info.data.borrow())?;
-    if obligation_info.owner != program_id {
+    let mut repay_reserve = Reserve::unpack(&repay_reserve_info.data.borrow())?;
+    if repay_reserve_info.owner != program_id {
+        msg!("Repay reserve provided is not owned by the lending program");
         return Err(LendingError::InvalidAccountOwner.into());
     }
-    if &obligation.borrow_reserve != repay_reserve_info.key {
-        msg!("Invalid repay reserve account");
+    if &repay_reserve.lending_market != lending_market_info.key {
+        msg!("Repay reserve lending market does not match the lending market provided");
         return Err(LendingError::InvalidAccountInput.into());
     }
-    if &obligation.collateral_reserve != withdraw_reserve_info.key {
-        msg!("Invalid withdraw reserve account");
+    if &repay_reserve.liquidity.supply_pubkey != repay_reserve_liquidity_supply_info.key {
+        msg!("Repay reserve liquidity supply does not match the repay reserve liquidity supply provided");
         return Err(LendingError::InvalidAccountInput.into());
     }
-    if obligation.deposited_collateral_tokens == 0 {
-        return Err(LendingError::ObligationEmpty.into());
-    }
-
-    let mut repay_reserve = Reserve::unpack(&repay_reserve_info.data.borrow())?;
-    if repay_reserve_info.owner != program_id {
-        return Err(LendingError::InvalidAccountOwner.into());
+    if &repay_reserve.liquidity.supply_pubkey == source_liquidity_info.key {
+        msg!("Repay reserve liquidity supply cannot be used as the source liquidity provided");
+        return Err(LendingError::InvalidAccountInput.into());
     }
-    if &repay_reserve.lending_market != lending_market_info.key {
-        msg!("Invalid reserve lending market account");
+    if &repay_reserve.collateral.supply_pubkey == destination_collateral_info.key {
+        msg!(
+            "Repay reserve collateral supply cannot be used as the destination collateral provided"
+        );
         return Err(LendingError::InvalidAccountInput.into());
     }
+    if repay_reserve.last_update.is_stale(clock.slot)? {
+        msg!("Repay reserve is stale and must be refreshed in the current slot");
+        return Err(LendingError::ReserveStale.into());
+    }
 
     let withdraw_reserve = Reserve::unpack(&withdraw_reserve_info.data.borrow())?;
     if withdraw_reserve_info.owner != program_id {
+        msg!("Withdraw reserve provided is not owned by the lending program");
         return Err(LendingError::InvalidAccountOwner.into());
     }
-    if withdraw_reserve.lending_market != repay_reserve.lending_market {
-        return Err(LendingError::LendingMarketMismatch.into());
-    }
-
-    if repay_reserve_info.key == withdraw_reserve_info.key {
```

Comment:
Yeah, this is a funny case. On the one hand, it seems kind of bad to let someone get out of their loan cheaper by liquidating themselves than by repaying, but on the other, they don't really want to take the risk that someone else liquidates them first... so I guess it's okay?

Reply:
Sorry, I'm not sure I follow how liquidating themselves would be cheaper than repaying, can you elaborate?

Reply:
Okay, so let's say they deposit SOL and borrow SOL against it. Their loan becomes unhealthy, but instead of repaying their borrowed SOL, they liquidate it for `liquidation_bonus` % less.

Reply:
Hmm I think it's only is cheaper in the end if the protocol loses value, which wouldn't necessarily happen here as long as the liquidation threshold is below 100%

Reply:
In any case, I think this is fine as is